federation/pkg/kubefed/init/init.go

```diff
@@ -358,29 +398,56 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
 		endpoint = endpoint + ":" + strconv.Itoa(int(svc.Spec.Ports[0].NodePort))
 	}
 
-	// 8. Write the federation API server endpoint info, credentials
-	// and context to kubeconfig
 	err = updateKubeconfig(config, i.commonOptions.Name, endpoint, i.commonOptions.Kubeconfig, entKeyPairs, i.options.dryRun)
 	if err != nil {
+		glog.V(4).Infof("Failed to update kubeconfig: %v", err)
 		return err
 	}
+	fmt.Fprintln(cmdOut, "done")
+	glog.V(4).Info("Successfully updated kubeconfig")
 
 	if !i.options.dryRun {
-		fedPods := []string{serverName, cmName}
-		err = waitForPods(hostClientset, fedPods, i.commonOptions.FederationSystemNamespace)
-		if err != nil {
-			return err
-		}
-		err = waitSrvHealthy(config, i.commonOptions.Name, i.commonOptions.Kubeconfig)
+		fmt.Fprint(cmdOut, "Waiting for federation control plane to come up...")
+		glog.V(4).Info("Waiting for federation control plane to come up")
+		err = runAndShowProgress(cmdOut, func() error {
+			fedPods := []string{serverName, cmName}
+			innerErr := waitForPods(hostClientset, fedPods, i.commonOptions.FederationSystemNamespace)
+			if innerErr != nil {
+				return innerErr
+			}
+			return waitSrvHealthy(config, i.commonOptions.Name, i.commonOptions.Kubeconfig)
+		})
+		glog.V(4).Info("Federation control plane running")
+		fmt.Fprintln(cmdOut, "done")
 		if err != nil {
 			return err
 		}
 		return printSuccess(cmdOut, ips, hostnames, svc)
 	}
-	_, err = fmt.Fprintf(cmdOut, "Federation control plane runs (dry run)\n")
+	_, err = fmt.Fprintln(cmdOut, "Federation control plane runs (dry run)")
+	glog.V(4).Info("Federation control plan runs (dry run)")
 	return err
 }
 
+// runAndShowProgress runs f in a goroutine, displaying a progress indicator
+// as it runs.
+func runAndShowProgress(cmdOut io.Writer, f func() error) error {
+	done := false
+	var err error
+	go func() {
+		err = f()
+		done = true
+	}()
+
+	return wait.Poll(2*time.Second, initStepTimeout, func() (bool, error) {
+		fmt.Fprint(cmdOut, ".")
+		if !done {
```

Comment:
The usage of done as a flag is racy (not the most unsafe) but i believe a channel would be cleaner.

Also, what's the implication of the timeout that this PR introduces? First, it would be good to document how 2 minutes is determined to be a safe default. Also, does this new timing out behavior affect kubefed's ability to clean up existing resources?